specs/PushNotifications/PushNotifications-spec.md

```diff
@@ -0,0 +1,378 @@
+* Supporting raw push for Packaged Win32(MSIX)
+* Supporting raw push for Unpackaged Win32.
+* Supporting Visual Cloud Toasts for packaged Win32 apps.
+* Supporting Visual Cloud Toasts for unpackaged Win32 apps.
+
+A Portal Registration flow through [AAD (Azure Active Directory)](https://docs.microsoft.com/en-us/azure/active-directory/develop/quickstart-register-app) will also be defined that decouples the dependence of the Push Flow with the Partner Centre Portal. 
+The following are the main steps in the Push flow with AAD:
+* The developer will first obtain a Remote Identifer GUID and a client secret as part of this Portal Registration flow. This is a one time operation by the developer to provision an app. 
+* When requesting a channel using WNS client Reunion APIs, the developer will pass in a remoteId and WNS will return a specially encoded Channel URL with that Remote Identifier. 
+* When the application service Pushes a notification down to the client, it will retrieve an access token with AAD using the client secret provisioned above. 
+* The application service will then pass that access token along with the payload in a Http POST operation to the channel URL. 
+* WNS will ensure that the payload is ultimately delivered to the app residing on the client device.
+
+Phase 1 (Reunion Preview):
+WNS Push APIs will only support WIN32 Packaged app (MSIX).
+
+Phase 2 (Reunion V1):
+Our Push APIs will add support for Unpackaged Win32 scenarios.
+
+# Examples
+
+## Registration of the Push Activator for Win32 (Inproc)
+If a Win32 app needs to subscribe to Push triggers inproc, the code in Main would follow the pattern below:
+* The Registration will take in a CLSID of the COM server as part of the Activator setup. The registration API will simply be a thin wrapper around the BackgroundInfra WinRT APIs and abstract away the COM details from the developer.
+* The app will query ApplifeCycle APIs to retrieve an ActivationKind. (Note: This is covered in a seperate AppLifeCycle API spec.)
+* If the Activation is Push, the app will QI the Arguments property to retrieve an instance of PushActivatedEventArgs and get the Push payload from it. Care needs to be taken to Take a Deferral and Release the Deferral while processing the payload for Power Management compliance.
+* If the Activation is Foreground, the app will do 2 things:
+  * It will subscribe to a Foreground In-memory Push event
+  * It will request a Push Channel Asynchronously with an implementation of Progress and Completed event handler.
+    * Expect Channel operations to take around 2-16 minutes in some rare cases (retryable errors). In the 90th percentile case, it should be fairly quick operation (in a few seconds).  
+
+```cpp
+int main()
+{
+    // Always Register the Push trigger as the first step: It sets up the Inproc COM server for subsequent Push operations
+    winrt::guid activatorGuid{ 0xC380465D, 0x2271, 0x428C, { 0x9B, 0x83, 0xEC, 0xEA, 0x3B, 0x4A, 0x85, 0xC1} };
```

Comment:
Is there possibly a way for us to not require providing an activator GUID? Making the API super simple? Having to create and register a GUID is a bit of a pain, maybe we can derive a unique GUID from their app identity or something...

Reply:
@DrusTheAxe With MSIX, I understand we have a way to parse the manifest and retrieve an Activator GUID. However I was told that there were significant perf implications with doing this. Any thoughts?

Reply:
That depends. By 'manifest' you mean `appxmanifest.xml`? What were you thinking of putting in there?

And who's manifest? One of ProjectReunion's packages? Or a packaged apps' manifest?

Are you thinking of a [SparsePackage](https://blogs.windows.com/windowsdeveloper/2019/10/29/identity-registration-and-activation-of-non-packaged-win32-apps/)? A hybrid in many ways but in the end still packaged thus has appxmanifest.xml.

What would a not-packaged app do? They have no appxmanifest.xml

One option that's not perf-bad is [custom properties in an appExtension](https://docs.microsoft.com/windows/msix/desktop/custom-props-app-extensions). The Windows.ApplicationModel.AppExtensions API gets you access to that reasonably efficiently. One caveat is that's an application-scope `<Extension>` so framework packages can't define them. And there's no package-scope equivalent (see [Support custom metadata field for optional packages (and possibly other type of packages) #226](https://github.com/microsoft/ProjectReunion/issues/226) for more details).

What you want to avoid is "Find, Load and Parse AppxManifest.xml" unless absolutely forced to (and even then, think again :P). The perf implications are significant, as we know from 1st hand experience (more than once...).

Reply:
@DrusTheAxe  A packaged apps manifest specifically is being inferred to here. In some of my earlier conversations with Anis and other folks, I was told that the cost of dynamically loading, parsing and retrieving the GUID was too high which is why this approach was recommended.

For non-packaged apps, we would be relying on other forms of Activation like Protocol Launch.